popstats/server.js

```diff
@@ -0,0 +1,268 @@
+  rpcThrottle((hash) => evmClient.getTransaction({ hash }))
+);
+
+// Database --------------------------------------------------------------------
+
+const keyPrefix = config.testnet
+  ? config.redis.prefixes.testnet
+  : config.redis.prefixes.mainnet;
+const bestBlockKey = `${keyPrefix}:${config.redis.bestBlockKey}`;
+
+const concurrencyLimit = pLimit(config.upstashLimits.concurrency);
+const dailyThrottle = pThrottle(config.upstashLimits.daily);
+const perSecondThrottle = pThrottle(config.upstashLimits.perSecond);
+
+const connectToRedis = () =>
+  createClient(config.redis)
+    .on("connect", function () {
+      beforeExit.do(() => this.disconnect());
+    })
+    .on("error", function (err) {
+      console.error("Redis error:", err.code);
+      process.exit(1);
+    })
+    .connect();
+
+const getRedisClient = dailyThrottle(perSecondThrottle(pMem(connectToRedis)));
+
+const getBestBlockNumber = () =>
+  concurrencyLimit(() =>
+    getRedisClient().then((redisClient) =>
+      redisClient
+        .get(bestBlockKey)
+        .then((bestBlock) => (bestBlock ? BigInt(bestBlock) : undefined))
+    )
+  );
+
+const setBestBlockNumber = (bestBlock) =>
+  concurrencyLimit(() =>
+    getRedisClient()
+      .then((redisClient) =>
+        redisClient.set(bestBlockKey, bestBlock.toString())
+      )
+      .then(function () {
+        console.log(`Best block set to ${bestBlock}`);
+      })
+  );
+
+// Absinthe --------------------------------------------------------------------
+
+function insertApiPoints(rows) {
+  console.log(`Inserting ${rows.length} rows`);
+  const headers = new Headers();
+  headers.append("Authorization", `Bearer ${config.absintheApi.key}`);
+  headers.append("Content-Type", "application/json");
+  const graphql = JSON.stringify({
+    query: `mutation ($objects: [api_points_insert_input!]!) {
+              insert_api_points(objects: $objects) {
+                returning { id }
+              }
+            }`,
+    variables: { objects: rows },
+  });
+  return fetch(/** @type {string} */ (config.absintheApi.url), {
+    body: graphql,
+    headers,
+    method: "POST",
+  })
+    .then((res) => res.json())
+    .then(function ({ errors }) {
+      if (errors) {
+        throw new Error(`Insertion failed: ${JSON.stringify(errors)}`);
+      }
+
+      console.log("Insertion done");
+    });
+}
+
+// Events processing -----------------------------------------------------------
+
+function getMoreGovernanceTokenTransfers({ fromBlock, toBlock, transfers }) {
+  const maxToBlock = BigIntMath.min(
+    fromBlock + config.blocks.maxWindow - 1n,
+    toBlock
+  );
+  console.log(`Getting transfers between blocks ${fromBlock}-${maxToBlock}`);
+  return getGovernanceTokenTransfers({
+    fromBlock,
+    toBlock: maxToBlock,
+  }).then((moreTransfers) => ({
+    fromBlock: maxToBlock + 1n,
+    toBlock,
+    transfers: transfers.concat(moreTransfers),
+  }));
+}
+
+const thereAreBlocksToProcess = ({ fromBlock, toBlock }) =>
```

Comment:
I think this reads better as a question, rather than an affirmative question 🤔  

```suggestion
const areThereBlocksToProcess = ({ fromBlock, toBlock }) =>
```

Thoughts?

Reply:
ok, now I see how it would read better:

"do whilst there are blocks to process"

nvm then!